src/routes/api/gasless/call/+server.ts

```diff
@@ -0,0 +1,358 @@
+      chains: [network.gqlName],
+    },
+    fetch,
+  );
+
+  if (!isProjectUnclaimedQueryResponse.projectById)
+    return error(400, '{ "error": "Project not found" }');
+  const chainData = filterCurrentChainData(isProjectUnclaimedQueryResponse.projectById.chainData);
+  if (chainData.__typename === 'ClaimedProjectData')
+    return error(400, '{ "error": "Project already claimed" }');
+
+  return true;
+}
+
+function validateCollectEarningsTx(txDescription: TransactionDescription): boolean {
+  if (txDescription.name !== 'callBatched') return false;
+
+  const parseCallsResult = parseCallBatchedArgs(txDescription);
+  if (!parseCallsResult.success) return false;
+
+  const parsedCalls = parseCallsResult.data;
+
+  // Ensure all calls are either
+  // 1. Collect on drips,
+  // 2. ReceiveStreams on address driver,
+  // 3. SqueezeStreams on address driver,
+  // 4. Split on address driver, or
+  // 4. Unwrap on native token unwrapper
+
+  for (const [target, calldata] of parsedCalls) {
+    const targetIsDrips = target.toLowerCase() === network.contracts.DRIPS.toLowerCase();
+    const targetIsNativeTokenUnwrapper =
+      target.toLowerCase() === network.contracts.NATIVE_TOKEN_UNWRAPPER?.toLowerCase();
+    const targetIsAddressDriver =
+      target.toLowerCase() === network.contracts.ADDRESS_DRIVER.toLowerCase();
+
+    if (targetIsDrips) {
+      const dripsIface = new Interface(dripsAbi);
+      const dripsParseRes = dripsIface.parseTransaction({ data: calldata });
+
+      if (!dripsParseRes) return false;
+
+      if (
+        dripsParseRes.name !== 'receiveStreams' &&
+        dripsParseRes.name !== 'squeezeStreams' &&
+        dripsParseRes.name !== 'split'
+      ) {
+        return false;
+      }
+    } else if (targetIsAddressDriver) {
+      const addressDriverIface = new Interface(addressDriverAbi);
+      const addressDriverParseRes = addressDriverIface.parseTransaction({ data: calldata });
+
+      if (!addressDriverParseRes) return false;
+
+      if (addressDriverParseRes.name !== 'collect') return false;
+    } else if (targetIsNativeTokenUnwrapper) {
+      const iface = new Interface(nativeTokenUnwrapperAbi);
+      const parseRes = iface.parseTransaction({ data: calldata });
+
+      if (!parseRes) return false;
+
+      if (parseRes.name !== 'unwrap') return false;
+    } else {
+      return false;
+    }
+  }
+
+  return true;
+}
+
+/** Verify that the calldata corresponds to a supported gasless action. */
+async function inferActionType(
+  calldata: string,
+  contractName: string,
+): Promise<SupportedGaslessAction | null> {
+  const contractAbi = Object.entries(SUPPORTED_CONTRACT_ABIS).find(
+    ([name]) => name === contractName,
+  )?.[1];
+  assert(contractAbi, 'Invalid contract name');
+
+  const iface = new Interface(contractAbi);
+
+  const parseRes = iface.parseTransaction({ data: calldata });
+  assert(parseRes, 'Invalid calldata');
+
+  switch (contractName) {
+    case 'CALLER': {
```

Comment:
Did you mean to add more cases here? This is just an `if` otherwise.

Reply:
actually, the entire logic is built with having multiple possible target contracts in the future (`SUPPORTED_CONTRACT_ABIS`, taking in `contractName` etc.), and while there's currently only Caller, I think that's fine. I think the likelihood that we want to make many more things gasless in the near future on cheap networks is pretty high. editing splits, updating metadata etc. would all go against either NFTDriver or RepoDriver, not caller, because they're not batches, so we'd need to add more contracts here.